metrics: Add cilium-health metrics about latency

vadorovsky opened the pull request:
This change adds the Prometheus metrics server for cilium-health
daemon.

Currently supported metric is a histogram with information about
latency to nodes and endpoints.

Fixes #4267

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4596)
<!-- Reviewable:end -->

joestringer commented on daemon/main.go:
```diff
@@ -432,6 +433,9 @@ func init() {
 	viper.BindEnv("prometheus-serve-addr-deprecated", "PROMETHEUS_SERVE_ADDR")
+	flags.StringVar(&healthPrometheusServeAddr,
+		"health-prometheus-serve-addr", "", "IP:Port on which to serve prometheus metrics from cilium-health (pass \":Port\" to bind on all interfaces, \"\" is off)")
+	viper.BindEnv("health-prometheus-serve-addr", "CILIUM_HEALTH_PROMETHEUS_SERVE_ADDR")
```
One thought I have is that if we moved this host-side instance of `cilium-health` out into its own sidecar, then we don't need to pass these health agent-specific bits of information through the cilium daemon.

At least initially, I'd expect the cilium-health instance run as an endpoint would remain how it is, launched by the daemon.

(I realise this is a bit orthogonal to what you're trying to achieve, but longer term I think it's probably the direction we want to go).

vadorovsky replied:
We still need to pass information through the cilium daemon, at least about Prometheus metrics. I mean, serving Prometheus metrics from endpoints is still needed. I didn't include that part of work in this PR yet, but I will need to do that. We need both host and endpoint latency stats in Prometheus

joestringer replied:
If we want to ensure there is only one prometheus port exposed, then sure. However the regular `cilium-health` instance that Cilium runs should be accessible via the same IP as Cilium (host IP as it's run in `hostNetworking` mode), so we could expose an additional prometheus port for the cilium-health agent. See also the discussion below on this PR.

vadorovsky commented on contrib/systemd/cilium-health:
```diff
@@ -0,0 +1 @@
+CILIUM_HEALTH_OPTS="-d --prometheus-serve-addr=:9091"
```
I'm wondering whether we should include Prometheus support by default or not. If not, then maybe we should provide some env variable for Vagrant, i.e. `PROMETHEUS` - if enabled, then our vagrant scripts should add `--prometheus-serve-addr` to sysconfig files.

@aanm @ianvernon @eloycoto WDYT?

vadorovsky commented on pkg/health/server/prober.go:
```diff
@@ -250,6 +251,13 @@ func (p *prober) httpProbe(node string, ip string, port int) *models.Connectivit
 			result.Status = ""
 			result.Latency = rtt.Nanoseconds()
+
+			metrics.Latency.WithLabelValues(
```
Those two places are the only places in this file which make requests and store the info about latency. Or am I wrong?

I'm struggling with understanding where we do latency probes to hosts, and where to endpoints.

@joestringer @raybejjani Any hints would be appreciated.

joestringer replied:
I don't think that there's any special logic for hosts vs. endpoints, only for ICMP vs. HTTP probes.

vadorovsky reviewed (commented):
@SuperQ @mdlayher Currently we are defining the port to scrape here:

https://github.com/cilium/cilium/blob/master/examples/kubernetes/cilium.yaml#L94

Is there any way to define multiple ports? We are going to serve cilium-health metrics on port 9091, from an another container.

vadorovsky commented on contrib/systemd/cilium:
```diff
@@ -18,11 +18,13 @@ PATH=/usr/lib/llvm-3.8/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sb
 # --kvstore consul
 # Set Key-value store options:
 # --kvstore-opt consul.address=127.0.0.1:8500
+# Set IP:Port on which to serve prometheus metrics:
```
@nirmoy @darix Is this format correct? PTAL

vadorovsky commented on contrib/systemd/cilium-health:
```diff
@@ -0,0 +1,24 @@
+### Configuration options for cilium-health
```
@nirmoy @darix Is this format correct? PTAL

nirmoy replied:
LGTM

aanm replied:
@mrostecki can you append `.service` to the filename?

joestringer commented on pkg/health/server/prober.go:
```diff
@@ -250,6 +251,13 @@ func (p *prober) httpProbe(node string, ip string, port int) *models.Connectivit
+				"",
+				ip,
+				"http",
+			).Observe(rtt.Seconds())
```
Awesome, looks like `Seconds()` should grab it in floating point which seems right, the metric can be standardized in seconds but we should have the same degree of accuracy as what we want.

How do you want to represent if the probe fails? latency=0? latency=math.MaxFloat64?

vadorovsky replied:
latency=0 would be probably the most convenient

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -32,8 +32,14 @@ import (
 )
 
+const (
+	healthSocket = "/run/cilium/prometheus_health.sock"
```
Does this need to be in the metrics package?

vadorovsky replied:
No, it's a leftover from my previous code, when I tried my crazy idea with proxying metrics server. Sorry, I will remove that.

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -206,57 +212,74 @@ var (
+	// Latency is a histogram showing latency to nodes
+	Latency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: Namespace,
+		Name:      "latency",
```
Please set this up, curl it and pass the results into `promtool check-metrics`. This will provide suggestions about naming.

At minimum it should state the units, and probably whether it's a sum or count.

If you haven't read up on the following pages, they may prove interesting for how to choose names and types for prometheus metrics:
https://prometheus.io/docs/practices/naming/
https://prometheus.io/docs/practices/histograms/

vadorovsky replied:
Thanks for the hint. I added `_seconds` suffix to the name and it's accepted by promtool.

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -206,57 +212,74 @@ var (
+		Namespace: Namespace,
+		Name:      "latency",
+		Help:      "Latency to nodes",
+	}, []string{"node", "type", "ip", "protocol"})
```
What do you plan for the "type" to indicate? I note that it's left empty currently in both usages.

vadorovsky replied:
I was thinking about putting "host" or "endpoint" as a value of "type".

joestringer replied:
I actually like this idea, I think it's an important differentiation.

joestringer commented on daemon/main.go:
```diff
@@ -87,35 +88,36 @@ var (
+	masquerade                bool
+	nat46prefix               string
+	prometheusServeAddr       string
+	healthPrometheusServeAddr string
```
I don't think this is used any more, this patch fragment can be dropped?

vadorovsky replied:
Yes, it can be dropped.

joestringer reviewed (commented):
Apologies about the delays in providing feedback, I came down quite sick last week. I provided some various feedback.

There's probably a step beyond this where the health endpoint can be built/distributed in a separate container but this LGTM. Haven't had a chance to spin it up and try it yet though.

joestringer commented on daemon/main.go:
```diff
@@ -753,10 +755,10 @@ func runDaemon() {
+	// Watch cilium-health daemon.
+	log.Info("Watching Cilium health daemon")
+	d.ciliumHealth = &healthwatch.CiliumHealth{}
+	go d.ciliumHealth.Watch()
```
I do wonder, now that the cilium-health instance that polls is being invoked as a separate container, whether it's useful to have the cilium daemon keep an eye on it, or if we could just drop this check and the related underlying code.

vadorovsky replied:
cilium-agent has some pieces of code which are looking for status of cilium-health. I didn't want to touch them, but if you think that we can just drop them, I can do that.

joestringer replied:
I'm on the fence about it, so it's probably simpler to just leave this in place unless someone voices a strong opinion either way.

One of the things I have some concern about is that someone upgrades Cilium from an earlier version without updating the `DaemonSet`, then the new Cilium doesn't run the `cilium-health` instance, but `cilium-agent` still relies on it, then the health checks for `cilium-agent` fail (is this true?), causing Cilium to get into a bad state. That said, this is clearly outside the scope of our upgrade instructions[0], so maybe it's fair to just tell users "well, if you don't follow the instructions, this is what happens".

[0] http://docs.cilium.io/en/latest/install/upgrade/

joestringer commented on pkg/health/server/prober.go:
```diff
@@ -398,6 +421,12 @@ func newProber(s *Server, nodes nodeMap) *prober {
+			node.Name,
+			addr.String(),
+			"icmp",
+		).Observe(rtt.Seconds())
```
I think we're still missing the failedProbeLatency case for ICMP?

vadorovsky replied:
Well, I don't see any error handling in this function. Under what condition should I add failedProbeLatency case there?

joestringer reviewed (changes requested):
> @joestringer in cilium-health status we actually differentiate hosts and endpoints. They're differentiated here:
https://github.com/cilium/cilium/blob/master/pkg/health/server/prober.go#L125
I'm wondering how I can do that distinction at the time of probe, when I want to save a metric.

We may need to consider adjusting the structures of how we store information about nodes / endpoints in the health server, for instance to include a bool for whether the IP represents an endpoint or not. FWIW here's the API definition for the node/endpoint information that `cilium-health` fetches from `cilium`:

https://github.com/cilium/cilium/blob/d6b61dcfe16d8764f716982b93dd3febd47ad358/api/v1/openapi.yaml#L1258-L1262

https://github.com/cilium/cilium/blob/d6b61dcfe16d8764f716982b93dd3febd47ad358/api/v1/openapi.yaml#L1379-L1394

I think that at the moment, this is stored in the `nodesMap` but we discard the information when generating the list of nodes in `getNodeAddresses()`, which is done just before performing a probe. Maybe the format for those can just be reworked a little to allow this information to be available for metrics gathering.

> @joestringer OK, I see that cilium-health status doesn't show endpoint info anymore. So I guess I can just remove the "type" field.

This is surprising to me. It should still show endpoint info and if not then that sounds like a bug. Can you reliably reproduce this behaviour? Does it also affect master?

vadorovsky committed:
doc: Add information about cilium-health service

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

aanm commented on cilium-health/cmd/root.go:
```diff
@@ -167,6 +170,13 @@ func runServer() {
 	}()
 
+	if prometheusServeAddr := viper.GetString("prometheus-serve-addr"); prometheusServeAddr != "" {
+		log.Info("Serving cilium-health prometheus metrics")
```
Print the address for which we are serving the metrics. `log.WithField(...`

aanm commented on cilium-health/launch/endpoint.go:
```diff
@@ -45,6 +46,8 @@ import (
 )
 
 var (
+	log = logging.DefaultLogger
```
cc @ianvernon was your PR with a dedicated name, the package name, for each logger merged?

vadorovsky replied:
Yes, it was. I will rebase.

aanm commented on pkg/health/server/prober.go:
```diff
@@ -229,17 +273,24 @@ func (p *prober) setNodes(nodes nodeMap) {
 	result := &models.ConnectivityStatus{}
 
-	host := fmt.Sprintf("http://%s:%d", ip, port)
+	host := fmt.Sprintf("http://%s:%d", addr, port)
```
since you are touching this code please replace this with https://golang.org/pkg/net/#JoinHostPort

vadorovsky replied:
TBH, I think that using `JoinHostPort` will be completely counterproductive here. Why? Because that function takes strings as arguments. We have instances of `net.IPAddr` and `int`. Also, we still need to append `http://` prefix. So usage of `JoinHostPort` in this line would look like:

```
host := fmt.Sprintf("http://%s", net.JoinHostPort(addr.String(), strconv.FormatInt(port, 10)))
```

I'm sorry, but IMO this looks completely ugly.

aanm commented on pkg/health/server/prober.go:
```diff
@@ -273,33 +345,33 @@ func (p *prober) runHTTPProbe() {
 				if status.HTTP.Status != "" {
 					log.WithFields(logrus.Fields{
-						logfields.NodeName: name,
+						logfields.NodeName: node.Name,
```
Use `node.String()`

vadorovsky replied:
`node` doesn't have a `String` method

vadorovsky replied:
Although I'm fine with creating a follow-up PR  when this one gets merged.

aanm commented on pkg/metrics/metrics.go:
```diff
@@ -254,63 +260,80 @@ var (
 		// via an HTTP server. "/metrics" is the usual endpoint for that.
-		http.Handle("/metrics", promhttp.HandlerFor(registry, promhttp.HandlerOpts{}))
-		log.WithError(http.ListenAndServe(addr, nil)).Warn("Cannot start metrics server on %s", addr)
+		mux := http.NewServeMux()
```
why the `http.NewServeMux()`?

vadorovsky replied:
Because we are running two separate Prometheus metric servers. We cannot use just the standard mux instance.

aanm commented on examples/kubernetes/templates/v1/cilium-ds.yaml.sed:
```diff
@@ -159,6 +159,44 @@ spec:
+          periodSeconds: 5
+        volumeMounts:
+          - name: cilium-run
+            mountPath: /var/run/cilium
```
`            readOnly: true`

vadorovsky replied:
Why? `cilium-health` deamon is actually going to write to socket files, isn't it?

aanm replied:
@mrostecki can you try it out? I think readOnly is applied to the directory and not to the files inside of it.

aanm commented on Vagrantfile:
```diff
@@ -67,7 +67,9 @@ sudo mkdir -p /etc/sysconfig
 sudo cp /home/vagrant/go/src/github.com/cilium/cilium/contrib/systemd/cilium-consul.service /lib/systemd/system
 sudo cp /home/vagrant/go/src/github.com/cilium/cilium/contrib/systemd/cilium-docker.service /lib/systemd/system
 sudo cp /home/vagrant/go/src/github.com/cilium/cilium/contrib/systemd/cilium-etcd.service /lib/systemd/system
+sudo cp /home/vagrant/go/src/github.com/cilium/cilium/contrib/systemd/cilium-health.service /lib/systemd/system
```
I believe that in the previous commit the file didn't have `.service` in the filename

vadorovsky replied:
I have a proof that it has :) https://github.com/cilium/cilium/pull/4596/files#diff-101273d12c622fcf417ca435ebcca97c There are two files, with and without `.service`. The file without is a file for `/etc/sysconfig`.

aanm commented on contrib/vagrant/start.sh:
```diff
@@ -280,7 +281,7 @@ function write_cilium_cfg() {
     filename="${4}"
 
-    cilium_options="--auto-ipv6-node-routes --debug-verbose flow"
+    cilium_options="--auto-ipv6-node-routes --debug-verbose flow --prometheus-serve-addr=:9090"
```
Don't you need to to the same for the cilium-health options?

vadorovsky replied:
I think I will just remove that option. Maybe we should consider adding some env variable (i.e. `PROMETHEUS`) for enabling Prometheus metrics in Vagrant environment, but not in this PR.

aanm commented on Documentation/cmdref/cilium-health.md:
```diff
@@ -22,6 +22,8 @@ cilium-health
   -D, --debug                          Enable debug messages
   -H, --host string                    URI to cilium-health server API
   -i, --interval uint                  Interval (in seconds) for periodic connectivity probes (default 60)
+      --log-driver stringSlice         Logging endpoints to use for example syslog, fluentd
```
Can you add this as a separate PR?

vadorovsky committed:
logging: Add logfield for metrics server address

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
metrics: Add cilium-health metrics about latency

This change adds the Prometheus metrics server for cilium-health
daemon.

Currently supported metric is a histogram with information about
latency to nodes and endpoints.

Fixes #4267

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
systemd: Add cilium-health service

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
examples/kubernetes: Add cilium-health container

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

vadorovsky committed:
vagrant: Add cilium-health service and sysconfig

Signed-off-by: Michal Rostecki <mrostecki@suse.de>
